sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala

```diff
@@ -198,7 +204,29 @@ case class TransformWithStateExec(
       getOutputRow(obj)
     }
     ImplicitGroupingKeyTracker.removeImplicitKey()
-    mappedIterator
+
+    // Wrapper to ensure that the implicit key is set when the methods on the iterator
+    // are called. Inside of processNewData, we use a GroupedIterator, so handleInputRows
+    // is only called once per key. As such, we don't strictly need to set/unset the
+    // implicit key on every call to next(); we just need to set it on the first call
+    // to hasNext and unset it after the last call to next. However, such an optimization
+    // relies on knowing this implementation detail of processNewData, so we set/unset the
+    // implicit key on every call to a mappedIterator's public methods to be safe.
+    new Iterator[InternalRow] {
+      override def hasNext: Boolean = {
+        ImplicitGroupingKeyTracker.setImplicitKey(keyObj)
```

Comment:
Is it possible to set this and remove this once for the iterator lifetime ?

Reply:
I updated the code to fix this. The one down-side is that if _we_ accidentally call `next()` before `hasNext`, we can get an internal exception. But since that is a bug on _our_ part, it's fine if we throw an internal error, because it really would be one.